src/core/utils/components/VirtualScroll.tsx

```diff
@@ -100,145 +124,188 @@ export type VirtualScrollProps = {
-          onScrollRef.current,
-        )
-      : scrollContainer.current.removeEventListener(
-          'scroll',
-          onScrollRef.current,
-        );
-  }, []);
-
-  // Add event listener to the scrollable container.
-  React.useLayoutEffect(() => {
-    removeScrollListener();
-    onScrollRef.current = updateVirtualScroll;
-    if (
+      if (!parentRef.current) {
+        return;
+      }
+      parentRef.current.style.transform = `translateY(${getTranslateValue(
+        childHeight.current.child,
+        childHeight.current.firstChild,
+        startIndex,
+      )}px)`;
+    }, [bufferSize, itemsLength]);
+
+    const onScroll = React.useCallback(() => {
+      updateVirtualScroll();
+    }, [updateVirtualScroll]);
+
+    const removeScrollListener = React.useCallback(() => {
+      if (!onScrollRef.current) {
+        return;
+      }
       !scrollContainer.current ||
       scrollContainer.current === parentRef.current?.ownerDocument.body
-    ) {
-      parentRef.current?.ownerDocument.addEventListener(
-        'scroll',
-        updateVirtualScroll,
-      );
-    } else {
-      scrollContainer.current.addEventListener('scroll', updateVirtualScroll);
-    }
-    return removeScrollListener;
-  }, [updateVirtualScroll, removeScrollListener]);
-
-  React.useLayoutEffect(() => {
-    updateVirtualScroll();
-  }, [scrollContainerHeight, itemsLength, updateVirtualScroll]);
-
-  return (
-    <div
-      style={{
-        overflow: 'hidden',
-        minHeight: itemsLength * childHeight.current,
-        width: '100%',
-        ...style,
-      }}
-      ref={ref}
-      {...rest}
-    >
+        ? parentRef.current?.ownerDocument.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          )
+        : scrollContainer.current.removeEventListener(
+            'scroll',
+            onScrollRef.current,
+          );
+    }, []);
+
+    // Add event listener to the scrollable container.
+    React.useLayoutEffect(() => {
+      removeScrollListener();
+      onScrollRef.current = onScroll;
+      if (
+        !scrollContainer.current ||
+        scrollContainer.current === parentRef.current?.ownerDocument.body
+      ) {
+        parentRef.current?.ownerDocument.addEventListener('scroll', onScroll);
+      } else {
+        scrollContainer.current.addEventListener('scroll', onScroll);
+      }
+      return removeScrollListener;
+    }, [onScroll, removeScrollListener]);
+
+    React.useLayoutEffect(() => {
+      if (!scrollContainerHeight) {
+        return;
+      }
+
+      if (scrollToIndex > 0 && scrollContainer.current) {
+        const scrollTop =
+          (scrollToIndex - 1) * childHeight.current.child +
+          childHeight.current.firstChild;
+        scrollContainer.current.scrollTo({ top: scrollTop });
+      } else {
+        updateVirtualScroll();
+      }
+      // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [scrollContainerHeight]);
+
+    return (
       <div
```

Comment:
So this is the biggest issue.. you know how it ended up for us with polymorphism in certain client apps...

Reply:
react-window has the same
![image](https://user-images.githubusercontent.com/9084735/153939608-16951141-e05a-4021-a8a6-668b41fab1f1.png)
https://react-window.vercel.app/#/api/FixedSizeList

Reply:
What if we expose a hook as an alternative to the `VirtualScroll` component?

```tsx
const { outerProps, innerProps } = useVirtualization();

return (
  <div {...outerProps}>
    <ul {...innerProps}>{...}</ul>
  </div>
);
```

Reply:
Added prop `as`..
Looks really bad on combobox, situation with menu and styles..
Can you help me define the correct typing for `parentProps` .. ?

Reply:
it's a weird approach you've taken where `as` prop stays in the main props but the rest of the props are in `parentProps`. Now you can't use all the utility types we made in #409

Hard to say how to get correct typings but you can get default div typings if you do this:
![image](https://user-images.githubusercontent.com/9084735/156840093-39664244-ed18-4dd1-bcb2-5bdd0d620130.png)

I still think the hook approach will be a lot simpler. And it's not too much work. See [`24e21df`](https://github.com/iTwin/iTwinUI-react/commit/24e21df3f30354978e2161479f3b4b191b7e165a?w=1).

Reply:
Default I know, but I need this to react also to any custom element...
I did this on purpose, keeping all the props going to the root element and having a separate prop for the parent.. So much simpler and hopefully avoiding bugs (same as had with `Button`)..
Just could not figure the typings :/

Reply:
I tried the hook without any additional changes, but it does not work. Because VS gets mounted before even Menu is opened. And then nothing recalculates, when menu is opened..
Can I do conditional hook somehow ( I think not ) or I will need to add some additional logic to VS component...

Reply:
conditional hooks can be emulated by created a wrapper component which is conditionally rendered
![image](https://user-images.githubusercontent.com/9084735/157104954-93d9e1c9-8f91-4ca4-9768-a14d6b3e5abf.png)

Reply:
added